### PR TITLE
FM-891-Authenticate-Supplier-And-Buyer-Dashboard-Link

### PR DESCRIPTION
**Pull request by tim-s-ccs**
- Add authorisation for when the user clicks on the link
- Create tests for the authenticate user buyer side
- Add tests for authentication on the supplier side

To do this I created a method in the controller which checks that the contract they're trying to view through the link belongs to them

### CODE REVIEW

File: app/controllers/facilities_management/beta/supplier/contracts_controller.rb [hunk 1]
```diff
@@ -44,6 +45,12 @@ def set_contract
         def set_procurement
           @procurement = Procurement.find(@contract.procurement.id)
         end
+
+        def authenticate_supplier
+          return if current_user.email == @contract.supplier.data['contact_email']
```
Comment (iceraluk) [inline app/controllers/facilities_management/beta/supplier/contracts_controller.rb:50]: here you will have to do the same thing but you will probably need to save the supplier contact_email onto the contract(ProcurementSupplier). or maybe it works with just defining a method in the ProcurementSupplier model that pulls the email from supplier.data['contact_email'] and then in ability.rb you will just add `can :manage, FacilitiesManagement::ProcurementSupplier, supplier_email: user.email`

File: app/controllers/facilities_management/beta/procurements/contracts_controller.rb [hunk 2]
```diff
@@ -93,6 +94,12 @@ def contract_params
                   :contract_end_date_yyyy
                 )
         end
+
+        def authenticate_buyer
+          return if current_user == @procurement.user
+
+          redirect_to not_permitted_path(service: request.path_parameters[:controller].split('/').first)
+        end
```
Comment (iceraluk) [inline app/controllers/facilities_management/beta/procurements/contracts_controller.rb:102]: best way to add object specific authorisation is in cancan: https://github.com/CanCanCommunity/cancancan
you will have to add can :manage, FacilitiesManagement::Procurement, user_id: user.id in ability.rb for fm buyers

then in here you can just call authorize! :manage, @procurement

File: app/controllers/facilities_management/beta/supplier/contracts_controller.rb [hunk 3]
```diff
@@ -44,6 +45,12 @@ def set_contract
...
+
+          redirect_to not_permitted_path(service: request.path_parameters[:controller].split('/').first)
+        end
```
Comment (iceraluk) [inline app/controllers/facilities_management/beta/supplier/contracts_controller.rb:53]: here you will have to do the same thing but you will probably need to save the supplier contact_email onto the contract(ProcurementSupplier). or maybe it works with just defining a method 'supplier_email' in the ProcurementSupplier model that pulls the email from supplier.data['contact_email'] and then in ability.rb you will just add can :manage, FacilitiesManagement::ProcurementSupplier, supplier_email: user.email

File: app/models/ability.rb [hunk 4]
```diff
@@ -12,7 +13,10 @@ def initialize(user)
       service_specific_auth(user) if user.has_role? :buyer
       can :read, FacilitiesManagement::Supplier if user.has_role?(:fm_access) && user.has_role?(:supplier)
     end
+    can :manage, FacilitiesManagement::Procurement, user_id: user.id if user.has_role? :fm_access
+    can :manage, FacilitiesManagement::ProcurementSupplier, supplier_email: user.email if user.has_role?(:fm_access) && user.has_role?(:supplier)
   end
+  # rubocop:enable Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
Comment (iceraluk) [inline app/models/ability.rb:19]: just to get rid of complexity here you might wanna add the new roles in service_specific_auth(user) for buyer then maybe create a fm_supplier_service_specific_auth method for supplier access and include that one in there

Comment (tim-s-ccs) [reply app/models/ability.rb:19]: If I do this then anyone who is ccs_employee will not have access even if they are already a buyer

Comment (tim-s-ccs) [reply app/models/ability.rb:19]: But i'll see what I can do

File: app/models/ability.rb [hunk 5]
```diff
@@ -6,11 +6,10 @@ def initialize(user)
     cannot :manage, :all
     if user.has_role? :ccs_admin
       can :manage, :all
-    elsif user.has_role? :ccs_employee
-      admin_tool_specific_auth(user)
     else
+      admin_tool_specific_auth(user) if user.has_role? :ccs_employee
```
Comment (iceraluk) [inline app/models/ability.rb:10]: great!

**Pull request merged**